free5gc-operator/controllers/smfdeployment_controller.go

```diff
@@ -0,0 +1,648 @@
+/*
+Copyright 2023 The Nephio Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"html/template"
+	"sort"
+	"time"
+
+	appsv1 "k8s.io/api/apps/v1"
+	apiv1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	resource "k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/go-logr/logr"
+	workloadv1alpha1 "github.com/nephio-project/api/nf_deployments/v1alpha1"
+)
+
+// SMFDeploymentReconciler reconciles a SMFDeployment object
+type SMFDeploymentReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+type SMFcfgStruct struct {
+	PFCP_IP string // N4
+	// N4_IP   string
+	// N11_IP  string
+}
+
+type SMFAnnotation struct {
+	Name      string `json:"name"`
+	Interface string `json:"interface"`
+	IP        string `json:"ip"`
+	Gateway   string `json:"gateway"`
+}
+
+// Check string -> int/string -uint16 conversion
+func getSMFResourceParams(smfSpec workloadv1alpha1.SMFDeploymentSpec) (int32, *apiv1.ResourceRequirements, error) {
+	// Placeholder for Capacity calculation. Reqiurce requirements houlw be calculated based on DL, UL.
+
+	// TODO: increase number of replicas based on NFDeployment.Capacity.MaxSessions
+	var replicas int32 = 1
+	var cpuLimit string
+	var cpuRequest string
+	var memoryLimit string
+	var memoryRequest string
+
+	// spec.Capacity.MaxSessions <1000 -> 128Mi
+	// spec.Capacity.MaxNFConnections <10 -> 128Mi
+	if smfSpec.Capacity.MaxSessions < 1000 && smfSpec.Capacity.MaxNFConnections < 10 {
```

Comment:
Looks good. Glad to see this here. Though, IMO this is the sort of thing that belongs in a kpt function not an operator. But that is subject to debate, it's fine to do it here.